### PR TITLE
fix testing

### PR DESCRIPTION
**Pull request by ravvnen**
Fix the failing unit and integration tests on the feature/distributed-demo branch.

### CODE REVIEW

File: servers/lib/test/unit/files.service.unit.spec.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      jest
+        .spyOn(filesService, "getLocalFiles")
+        .mockResolvedValue(testDirectory);
+
+      const result = await filesService.Wrapper(pathToTestDirectory);
+
+      expect(result).toEqual(testDirectory);
+    });
+
+    it("should call getGitlabFiles when mode is gitlab", async () => {
```
[the remark this answers is not in the document]
Comment (ravvnen) [reply servers/lib/test/unit/files.service.unit.spec.ts:91]: @prasadtalasila  - again, this isnt anything that i want to refactor, as it defeats the purpose of readability in the test cases

**Comment by prasadtalasila** [comment]
@ravvnen , two comments:
1. Please rebase to resolve the conflicts
2. Do you have any idea about the issue with codeclimate and servers/lib problem? Can you check with other groups to see if this is unique to yours or @Artin13 also has the same problem with their python auth microservice?

File: docs/LIB-MS.md [hunk 2]
```diff
@@ -62,24 +62,27 @@ yarn install   # Install the required dependencies
...
-LOCAL_PATH='/home/dtaas'
+PORT='4001'
+MODE='local' or 'gitlab'
+LOCAL_PATH ='/Users/<Username>/DTaaS/files'
+GITLAB_GROUP ='dtaas'
 GITLAB_URL='https://gitlab.com/api/graphql'
 TOKEN='123-sample-token'
-GITLAB_GROUP='dtaas'
+LOG_LEVEL='debug'
+TEST_PATH='/Users/<Username>/DTaaS/servers/lib/test/data/test_assets'
```
Comment (prasadtalasila) [inline docs/LIB-MS.md:75]: This is no longer needed.

File: servers/config/gateway/dynamic/fileConfig.yml
Comment (prasadtalasila) [inline servers/config/gateway/dynamic/fileConfig.yml:1]: There is no need to change this file. The [existing one](https://github.com/INTO-CPS-Association/DTaaS/blob/feature/distributed-demo/servers/config/gateway/dynamic/fileConfig.yml) already has configuration for lib microservice.

**Review by prasadtalasila — COMMENTED** [review]
please fix the merge conflicts.

There seems to be some syntax errors in the code. Please see below.
```
$ script/syntax.bash
warning package.json: License should be a valid SPDX license expression
Running eslint
/home/vagrant/DTaaS/servers/lib/test/unit/files.service.unit.spec.ts
  62:42  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any   

✖ 1 problem (0 errors, 1 warning)

Done in 6.82s.
```

**Comment by prasadtalasila** [comment]
```
➜  lib git:(pr-62) ✗ ls test/data/test_assets/user2
 Test-Data2  'Test-Digital Twins'   Test-Functions   Test-Models   Test-Tools
➜  lib git:(pr-62) ✗ yarn test -i
yarn run v1.22.19
warning package.json: License should be a valid SPDX license expression
$ script/test.bash -i
warning package.json: License should be a valid SPDX license expression
 PASS  test/integration/files.service.integration.spec.ts (6.916 s)
  Integration Tests
    ✓ ensure that the getFiles method of the FilesService class returns the expected array of file names when called with a specific path (767 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        7.015 s
Ran all test suites matching /..\/test\/integration/i.
Done in 8.34s.
```

Is this expected?

**Comment by prasadtalasila** [comment]
During our last interaction you mentioned that tests uses the `server/lib/.env` file. That would not be a desirable situation. Please put in the required configuration in `test/data/` directory. Thanks.

**Comment by ravvnen** [comment]
> ```
> ➜  lib git:(pr-62) ✗ cat .env    
> PORT='4001'
> MODE='gitlab'
> LOCAL_PATH ='/home/vagrant/DTaaS/files'    
> GITLAB_GROUP ='dtaas'
> GITLAB_URL='https://gitlab.com/api/graphql'
> TOKEN='123-sample-token'
> LOG_LEVEL='debug'
> APOLLO_PATH='/lib'
> GRAPHQL_PLAYGROUND='true'
> 
> ➜  lib git:(pr-62) ✗ ls test/data/test_assets/user2
>  Test-Data2  'Test-Digital Twins'   Test-Functions   Test-Models   Test-Tools
> 
> ➜  lib git:(pr-62) ✗ yarn test -i
> yarn run v1.22.19
> warning package.json: License should be a valid SPDX license expression
> $ script/test.bash -i
> warning package.json: License should be a valid SPDX license expression
>  PASS  test/integration/files.service.integration.spec.ts (6.916 s)
>   Integration Tests
>     ✓ ensure that the getFiles method of the FilesService class returns the expected array of file names when called with a specific path (767 ms)
> 
> Test Suites: 1 passed, 1 total
> Tests:       1 passed, 1 total
> Snapshots:   0 total
> Time:        7.015 s
> Ran all test suites matching /..\/test\/integration/i.
> Done in 8.34s.
> ```
> 
> The result does not change even if the mode is changed to gitlab. Is this expected?

Yes i have set it up to yield the same answer, to simulate the monotonous answer

File: servers/config/gateway/README.md [hunk 3]
```diff
@@ -27,6 +28,23 @@ docker run -d \
 traefik:v2.5
 ```
 
+for mac
```
Comment (prasadtalasila) [inline servers/config/gateway/README.md:31]: please remove the mac-specific instructions

**Comment by prasadtalasila** [comment]
please update line-49 of .github/server.yml to "Upload test coverage to Codecov"

**Comment by prasadtalasila** [comment]
```
yarn install
yarn build #tests don't work without this step
yarn test -a
```
But this problem doesn't exist in PR #65 .

**Comment by prasadtalasila** [comment]
@ravvnen 
1. There are changes in `files/` directory.
2. The tests expect a gitlab repository with a certain file structure. Please add a `test/README.md` mentioning about these requirements.

**Comment by ravvnen** [comment]
> 2\. The tests expect a gitlab repository with a certain file structure. Please add a `test/README.md` mentioning about these requirements.

Good idea - I have created this README now, and it should document the necessary information.

File: files/user2/Data/.gitkeep
Comment (prasadtalasila) [inline files/user2/Data/.gitkeep:1]: These shouldn't have been renamed. The idea was to keep the previous names. The names should be lower case letters.

File: servers/config/gateway/README.md [hunk 4]
```diff
@@ -27,6 +27,10 @@ docker run -d \
 traefik:v2.5
 ```
 
+## Access
+
+# Run Lib MS at port 3000
```
Comment (prasadtalasila) [inline servers/config/gateway/README.md:32]: This is supposed to be 4001

File: servers/lib/test/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+Since the tests are tested across two different modes, 'local' or 'gitlab', a homogenous file structure is required.
+
+This file structure must be within 'files' aswell as within your working gitlab repository, as depicted below...
+
+local file system:
+
+```
+files/
+  user2/
+    Data/
```
Comment (prasadtalasila) [inline servers/lib/test/README.md:14]: why can't all these be lower case. If you absolutely need these to be upper case, then you need to update script/test.bash to do the task of changing the directory names in files/user2 and then reverting the structure at the end of the integration test.

File: servers/lib/test/e2e/app.e2e.spec.ts [hunk 6]
```diff
@@ -2,32 +2,31 @@ import { Test, TestingModule } from "@nestjs/testing";
 import { INestApplication } from "@nestjs/common";
 import * as request from "supertest";
 import { AppModule } from "../../src/app.module";
+import { e2eDirectory, pathToTestDirectory, testDirectory } from "../testUtil";
```
Comment (prasadtalasila) [inline servers/lib/test/e2e/app.e2e.spec.ts:5]: please take care of this warning.

File: servers/lib/test/README.md (hunk 5 above)
Comment (ravvnen) [reply servers/lib/test/README.md]: There is no necessary reason, so i have changed it all to lowercase - sorry about that

**Comment by prasadtalasila** [comment]
@ravvnen 
The files/user2 still have [upper case letters](https://github.com/INTO-CPS-Association/DTaaS/pull/62/files). I am wondering how the integration and e2e tests passed.

**Comment by ravvnen** [comment]
@prasadtalasila - i changed it to all lower-case 2 days ago, not sure how it is still upper case in there? I have .gitkeep in every folder ..

image

**Comment by prasadtalasila** [comment]
@ravvnen , I am pretty sure we talked about this issue, but can't remember why the following error comes.
```bash
➜  lib git:(pr-62) ✗ yarn test -a                                                                                                                                                      
yarn run v1.22.19                                                                                                                                                                      
warning package.json: License should be a valid SPDX license expression                                                                                                                
$ script/test.bash -a                                                                                                                                                                  
warning package.json: License should be a valid SPDX license expression                                                                                                                
[BABEL] Note: The code generator has deoptimised the styling of /home/vagrant/DTaaS/servers/lib/node_modules/lodash/lodash.js as it exceeds the max of 500KB.                          
[BABEL] Note: The code generator has deoptimised the styling of /home/vagrant/DTaaS/servers/lib/node_modules/lodash/lodash.js as it exceeds the max of 500KB.                          
 PASS  test/unit/files.service.unit.spec.ts (21.409 s)                                                                                                                                 
 PASS  test/unit/files.resolver.unit.spec.ts (33.634 s)                                                                                                                                
 FAIL  test/e2e/app.e2e.spec.ts (66.132 s)                                                                                                                                             
  ● End to End test for the application › should return the filename corresponding to the directory given in the query                                                                 
                                                                                                                                                                                       
    thrown: "Exceeded timeout of 10000 ms for a hook.                                                                                                                                  
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."                                     
                                                                                                                                                                                       
       8 |   let app: INestApplication;                                                                                                                                                
       9 |                                                                                                                                                                             
    > 10 |   beforeAll(async () => {                                                                                                                                                   
         |   ^                            
      11 |     const moduleFixture: TestingModule = await Test.createTestingModule({                                                                                           [94/261]
      12 |       imports: [AppModule],
      13 |     }).compile();

      at test/e2e/app.e2e.spec.ts:10:3
      at Object.<anonymous> (test/e2e/app.e2e.spec.ts:7:1)

 FAIL  test/integration/files.service.integration.spec.ts (66.215 s)
  ● Integration Tests › ensure that the getFiles method of the FilesService class returns the expected array of file names when called with a specific path

    thrown: "Exceeded timeout of 10000 ms for a hook.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      16 |   let filesService: FilesService;
      17 |
    > 18 |   beforeEach(async () => {
         |   ^
      19 |     const moduleFixture: TestingModule = await Test.createTestingModule({
      20 |       imports: [
      21 |         GraphQLModule.forRoot(getApolloDriverConfig()), // use your function

      at test/integration/files.service.integration.spec.ts:18:3
      at Object.<anonymous> (test/integration/files.service.integration.spec.ts:14:1)                                                                                          [72/261]


ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From test/e2e/app.e2e.spec.ts.

      13 |     }).compile();
      14 |     app = moduleFixture.createNestApplication();
    > 15 |     await app.listen(process.env.PORT);
         |     ^
      16 |   }, 10000);
      17 |   afterAll(async () => {
      18 |     await app.close();

      at node_modules/@nestjs/apollo/dist/drivers/apollo-base.driver.js:72:119
      at loadPackage (node_modules/@nestjs/common/utils/load-package.util.js:12:23)
      at ApolloDriver.registerExpress (node_modules/@nestjs/apollo/dist/drivers/apollo-base.driver.js:72:70)
      at ApolloDriver.registerServer (node_modules/@nestjs/apollo/dist/drivers/apollo.driver.js:38:24)
      at ApolloDriver.start (node_modules/@nestjs/apollo/dist/drivers/apollo.driver.js:23:20)
      at GraphQLModule.onModuleInit (node_modules/@nestjs/graphql/dist/graphql.module.js:105:9)
      at callModuleInitHook (node_modules/@nestjs/core/hooks/on-module-init.hook.js:43:5)
      at Proxy.callInitHook (node_modules/@nestjs/core/nest-application-context.js:228:7)
      at Proxy.init (node_modules/@nestjs/core/nest-application.js:107:5)
      at Proxy.init (node_modules/@nestjs/core/nest-application.js:107:5)                                                                                                      [51/261]
      at Proxy.listen (node_modules/@nestjs/core/nest-application.js:181:29)
      at Object.<anonymous> (test/e2e/app.e2e.spec.ts:15:5)

ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From test/integration/files.service.integration.spec.ts.

      26 |
      27 |     app = moduleFixture.createNestApplication();
    > 28 |     await app.init();
         |     ^
      29 |     filesService = moduleFixture.get<FilesService>(FilesService);
      30 |   }, 10000);
      31 |

      at node_modules/@nestjs/apollo/dist/drivers/apollo-base.driver.js:72:119
      at loadPackage (node_modules/@nestjs/common/utils/load-package.util.js:12:23)
      at ApolloDriver.registerExpress (node_modules/@nestjs/apollo/dist/drivers/apollo-base.driver.js:72:70)
      at ApolloDriver.registerServer (node_modules/@nestjs/apollo/dist/drivers/apollo.driver.js:38:24)
      at ApolloDriver.start (node_modules/@nestjs/apollo/dist/drivers/apollo.driver.js:23:20)
      at GraphQLModule.onModuleInit (node_modules/@nestjs/graphql/dist/graphql.module.js:105:9)
      at callModuleInitHook (node_modules/@nestjs/core/hooks/on-module-init.hook.js:43:5)
      at Proxy.callInitHook (node_modules/@nestjs/core/nest-application-context.js:228:7)
      at Proxy.callInitHook (node_modules/@nestjs/core/nest-application-context.js:228:7)
      at Proxy.init (node_modules/@nestjs/core/nest-application.js:107:5)
      at Object.<anonymous> (test/integration/files.service.integration.spec.ts:28:5)
--------------------|---------|----------|---------|---------|-------------------
File                | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
--------------------|---------|----------|---------|---------|-------------------
All files           |   96.15 |    88.46 |     100 |   95.74 |                   
 lib                |     100 |      100 |     100 |     100 |                   
  util.ts           |     100 |      100 |     100 |     100 |                   
 lib/src            |     100 |      100 |     100 |     100 |                   
  queries.ts        |     100 |      100 |     100 |     100 |                   
 lib/src/files      |   95.65 |    88.46 |     100 |   95.12 |                   
  files.resolver.ts |     100 |      100 |     100 |     100 |                   
  files.service.ts  |   94.44 |    88.46 |     100 |   93.93 | 18,22             
--------------------|---------|----------|---------|---------|-------------------
Test Suites: 2 failed, 2 passed, 4 total
Tests:       2 failed, 12 passed, 14 total
Snapshots:   0 total
Time:        67.251 s
Ran all test suites.
```

The same result is seen on both my computer and on integration server, but not on the GitHub actions. Why does it pass on github actions and not on other computers? Thanks.

**Comment by ravvnen** [comment]
Ahh yes, it was changing the beforeEach to beforeAll, which i have done now for both integration and e2e.

Please try it with this change. I am receiving no errors on github, linux or mac. Let me know if this change has fixed it.

**Pull request merged**